drivers: eth: e1000: Add driver Intel PRO/1000 Gigabit Ethernet contr…

ozhuraki opened the pull request:
…oller

This patch adds a driver for Intel PRO/1000 Gigabit Ethernet controller.

The driver currently supports only a single instance of the NIC.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

tbursztyka commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,340 @@
+#include <net/ethernet.h>
+#include <pci/pci.h>
+
+#define D(fmt, args...)	do {			\
```
From here :

tbursztyka commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,340 @@
+	D("R %s 0x%08x", e1000_reg_to_string(_reg), val); 	\
+	val;							\
+})
+
```
to here: place this into a eth_e1000_priv.h

tbursztyka commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,340 @@
+		&e1000_api,
+		ETHERNET_L2,
+		NET_L2_GET_CTX_TYPE(ETHERNET_L2),
+		E1000_MTU);
```
Always place the NET_DEVICE_INIT macro at the EOF (and thus remove the declaration od e1000_init in the beginning of the file.)

ozhuraki replied:
Since eth_e1000 is instantiated in the NET_DEVICE_INIT() macro, the IRQ_CONNECT() should be done below it. This would require one more function below NET_DEVICE_INIT() and a declaration for it. So, at the end NET_DEVICE_INIT() won't be at the end of file. I suggest to leave it as is.

tbursztyka replied:
Change it, there is a good reason I am asking this: we are going to replace this macro style stuff by built-time generated code at some point, and all drivers doing irq connect as you do (a minor amount of drivers do that yes) are going to give use more work.

tbursztyka commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,340 @@
+	union {
+		u64_t qw;
+		u32_t dw[2];
+	} addr = { .qw = dev->rx.addr };
```
We don't allow declaring a variable in the middle of a code block, so move that above

tbursztyka commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,340 @@
+
+	size_t len = e1000_linearize(pkt, dev->txb, sizeof(dev->txb));
+
+	net_pkt_unref(pkt);
```
The problem, currently, is that you driver should unref the packet only on success. And that you know only after your tx routine.

ozhuraki replied:
Good catch.

tbursztyka commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,340 @@
+	} addr = { .qw = dev->rx.addr };
+
+	if (!net_pkt_append_all(pkt, dev->rx.len - 4,
+			INT_TO_POINTER(addr.dw[0]), K_NO_WAIT)) {
```
indentation: should be below last '(' + 1 space. You can break a line before K_NO_WAIT if it's too long again

tbursztyka reviewed (commented):
tiny coments

galak commented on drivers/ethernet/Kconfig.e1000:
```diff
@@ -0,0 +1,21 @@
+	help
+	  Enable Intel(R) PRO/1000 Gigabit Ethernet driver.
+
+config ETH_E1000_IRQ
```
Add dt support, and get this property from there.

ozhuraki replied:
E1000 is a PCI only device and everything is discoverable through a PCI scan (including the IRQ number).

However, due to limitations of the IRQ_CONNECT() macro, where all parameters should be compile time constants, the IRQ line presently has to be configurable though Kconfig.

I briefly looked into device tree and can’t see any PCI or Ethernet support or similar use cases.

Can you be more specific with what kind of node, behind what bus, etc, should be defined in the device tree?

At the end, this still will need to be configurable (we would like to support a few other emulators in addition to qemu), unless the limitations of IRQ_CONNECT() are removed.

galak reviewed (changes requested):
have this driver use device tree and get the IRQ # from there.

jukkar reviewed (approved):
LGTM

tbursztyka commented on arch/x86/soc/ia32/soc.c:
```diff
@@ -26,4 +26,7 @@ MMU_BOOT_REGION(CONFIG_IOAPIC_BASE_ADDRESS, MB(1), MMU_ENTRY_WRITE);
 MMU_BOOT_REGION(CONFIG_HPET_TIMER_BASE_ADDRESS, KB(4), MMU_ENTRY_WRITE);
 #endif
+
+MMU_BOOT_REGION(0xfebc0000, KB(128), MMU_ENTRY_WRITE);
```
What is this this address? have CONFIG_ for it.

ozhuraki replied:
Fixed.

Commit:
dts/bindings: Add YAML bindings for the E1000 Ethernet controller

This is a very common Intel Ethernet controller found in various
targets.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add a dedicated config option for Ethernet DTS based entries

This will be usefull to tell when Ethernet device has its attributes
filled in by DTS rather than by Kconfig.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards/qemu_x86: Add a dts entry for the e1000 Ethernet controller

With approriate fixups.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>
Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
drivers: eth: e1000: Add driver for Intel PRO/1000 Ethernet controller

This patch adds a driver for Intel PRO/1000 Gigabit Ethernet controller.

The driver currently supports only a single instance of the NIC.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

tbursztyka commented:
@galak can you re-review please?

pfalcon commented on arch/x86/soc/ia32/soc.c:
```diff
@@ -26,4 +26,9 @@ MMU_BOOT_REGION(CONFIG_IOAPIC_BASE_ADDRESS, MB(1), MMU_ENTRY_WRITE);
 #endif
+
+#ifdef CONFIG_ETH_E1000
+MMU_BOOT_REGION(CONFIG_ETH_E1000_BASE_ADDRESS, KB(128), MMU_ENTRY_WRITE);
```
DTS [here](https://github.com/zephyrproject-rtos/zephyr/pull/9112/commits/6e7b16c1e47ffb65b22105e2826155146ef012b4#diff-8c50a13bfd9cee1a815504f63ecc98e0R34) specifies that there're just 0x100 bytes of registers, while here 128K is reserved. Is this correct?

pfalcon replied:
Also, ideally this would have been a separate commit...

pfalcon commented on drivers/ethernet/Kconfig.e1000:
```diff
@@ -0,0 +1,28 @@
+menuconfig ETH_E1000
+	bool
+	prompt "Intel(R) PRO/1000 Gigabit Ethernet driver"
+	depends on NET_L2_ETHERNET PCI_ENUMERATION
```
Is this correct? I don't remember ever seeing such syntax. However, it's easy to see following syntaxes:

~~~
timer/Kconfig:	depends on SOC_FAMILY_NRF && CLOCK_CONTROL_NRF5
~~~

and

~~~
Kconfig.nrfx:

config SPI_0_NRF_SPIS
        bool "nRF SPIS 0"
        depends on HAS_HW_NRF_SPIS0
        depends on SPI_SLAVE
~~~

I may imagine when options depends on 2 completely unrelated options, like here (Ethernet vs PCI), it's more readable to have 2 separated "depends on" lines, like the last example.

Calling @ulfalizer for suggestions.

pfalcon commented on drivers/ethernet/Kconfig.e1000:
```diff
@@ -0,0 +1,28 @@
+config ETH_E1000_BASE_ADDRESS
+	hex "Intel(R) PRO/1000 Gigabit Ethernet Base Address"
+	depends on ETH_E1000
+	default 0XFEBC0000
```
If we standardize on DTS, is this default still needed/useful?

ozhuraki replied:
“depends on” is fixed. There’s nothing wrong with the macros here. Register names are mostly 3 characters long, using READ/WRITE would visually dominate over the register names.

galak replied:
I agree w/Paul, we do we need to support non-DTS Kconfig?

pfalcon commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,252 @@
+	_(TDT);
+	_(RAL);
+	_(RAH);
+	}
```
Add `#undef _` here, to avoid unpleasant surprises in the future?

ozhuraki replied:
Fixed.

pfalcon commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,252 @@
+		ETHERNET_LINK_1000BASE_T;
+}
+
+static size_t e1000_linearize(struct net_pkt *pkt, void *buf, size_t bufsize)
```
Are you sure that we *still* don't have a reusable function which does that? ;-)

pfalcon commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,252 @@
+	dev->tx.len = data_len;
+	dev->tx.cmd = TDESC_EOP | TDESC_RS;
+
+	W(dev, TDT, 1);
```
What is "W"? Please use descriptive names.

pfalcon commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,252 @@
+		k_yield();
+	}
+
+	D("tx.sta: 0x%02hx", dev->tx.sta);
```
What is "D"? Please use descriptive names. And if there's a function named "do_foo_in_bar_way", it doesn't make sense to rename it to "d", please just use do_foo_in_bar_way() directly.

pfalcon commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,252 @@
+
+	D("tx.sta: 0x%02hx", dev->tx.sta);
+
+	return (dev->tx.sta & TDESC_STA_DD) ? E1000_SUCCESS : -EIO;
```
And here, the opposite situation. Why do we need E1000_SUCCESS? It's well-known convention that negative returns mean error, non-negative means success. And default non-negative value is 0, could just use it directly.

pfalcon commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -0,0 +1,252 @@
+static struct net_pkt *e1000_rx(struct e1000_dev *dev)
+{
+	struct net_pkt *pkt = NULL;
+	union {
```
If you never use dw[1], why these complications with union? Just cast u64_t to u32_t where needed.

ozhuraki replied:
Using a union is the standard way to do type punning.

pfalcon replied:
> Using a union is the standard way to do type punning.

Yes, if you need to go e.g. from float to its bits, as int's. Here again, just a narrowing cast from u64_t to u32_t would be enough. Up to you whether you want to change it, mostly a note.

ozhuraki replied:
> Up to you whether you want to change it, mostly a note.

I had another look at it, and dropped that union.

pfalcon reviewed (changes requested):
Looks mostly good and useful, thanks. Some suggestions though.

IMHO, at least matter of "depends on" and usage of D/E/R/W macros should be addressed, so let me give -1 so they weren't forgotten.

pfalcon commented:
Also, would be nice to get info on how to actually use this with qemu_x86. For starters, this info in quick form can go into a commit message, but ideally, it should be documented in Zephyr docs, as it will be important feature everyone will want to use.

jukkar commented on drivers/ethernet/Kconfig.e1000:
```diff
@@ -9,7 +9,7 @@
 	bool
 	prompt "Intel(R) PRO/1000 Gigabit Ethernet driver"
-	depends on NET_L2_ETHERNET PCI_ENUMERATION
+	depends on NET_L2_ETHERNET && PCI_ENUMERATION
```
Please squash this `drivers: eth: e1000: Update to new kconfig syntax` commit to `drivers: eth: e1000: Add driver Intel PRO/1000 Ethernet controller`

jukkar commented on drivers/ethernet/Kconfig.e1000:
```diff
@@ -13,6 +13,8 @@ menuconfig ETH_E1000
 	help
```
Please squash this commit `drivers: eth: e1000: Add IRQ priority and flags entries` to `drivers: eth: e1000: Add driver Intel PRO/1000 Ethernet controller`

pfalcon commented:
> “depends on” is fixed.

Thanks.

> There’s nothing wrong with the macros here. 
> Register names are mostly 3 characters long, using READ/WRITE would visually dominate over the register names.

Well, we don't control readability of register names in vendor hardwares. But we control readability of Zephyr source code. So, please just follow  conventions as used by other drivers:

~~~
$ zephyr/drivers/ethernet$ grep read *.c
eth_dw.c:	int_status = eth_read(context->base_addr, REG_ADDR_STATUS);
eth_enc28j60.c:		eth_enc28j60_read_reg(dev, ENC28J60_REG_MISTAT,
eth_sam_gmac.c:	i2c_burst_read_addr(dev, CONFIG_ETH_SAM_GMAC_MAC_I2C_SLAVE_ADDRESS,
phy_sam_gmac.c:	retval = phy_read(phy, MII_ANAR, &ability_adv);
~~~

e1000_read/e1000_write would be good names to use.

Related suggestion of using SYS_LOG_ERR() and friends directly also remains.

ozhuraki commented:
> So, please just follow conventions as used by other drivers
I can’t see any convention. Using functions here is technically worse, since we would lose the context in diagnostic output. Try to actually run the code and observe the difference with your proposal.

pfalcon commented:
> I can’t see any convention.

It's what I already said: the names are/should be descriptive.

> Using functions here is technically worse, since we would lose the context in diagnostic output.

Ah, that's not what I mean. Those don't have to be functions. And if you want to name them in caps to follow the standard macro convention, that's fine. As long as they aren't named "D" for debug, "W" for warning... No, wait, "W" means "write"! Well, that's exactly what I mean re: descriptive naming.

ozhuraki commented:
> [...] No, wait, "W" means "write"!

OK, addressed this. There's no such problem now.

pfalcon commented:
@ozhuraki:

> > 

    [...] No, wait, "W" means "write"!

> OK, addressed this. There's no such problem now.

Better, better, thanks! Now only `D` and `E` are left to go ;-).

Let me reiterate again on arguments, though most of them were already given: 

1. `D` and `E` are too short names and thus confusing (definitely will be for beginners). 
2. There's also problem with searching - with too short identifiers, there's usually a bunch of false positives.
3. But mot importantly, there's SYS_LOG_DBG and friends, which can be used directly. And most other drivers use those. (I write "most", but I actually didn't see a counter-example.)
4. If we let those in, we open a road for "in my special way" for individual programmers. For example, you like D and E, programmer1 likes DBG and ERR, programmer2 - DBG and ER (sic), programmer3 - DGB and ERR, and I like ABC and CDF (I like it weird).
5. Again, using standard API is useful for searching, grepping for "SYS_LOG_" will give any places where logging happens.


And well, I don't want this to look as personal nitpicking - @jukkar, you approved this PR, what do you think about usage of adhoc macros like D and E instead of standard API?

jukkar commented:
> what do you think about usage of adhoc macros like D and E instead of standard API?

The other ethernet drivers seem to be using SYS_LOG_xxx() macros so it might makes sense do the same here too. Those syslog macros will be converted to new logging macros anyway pretty soon so I do not see here a big issue with the printing. I would like to see this merged quite soon after 1.14 opens as this is a useful driver to have.

pfalcon commented:
> I would like to see this merged quite soon after 1.14 opens as this is a useful driver to have.

Me too, so I really hope that @ozhuraki would listen to suggestions on how to improve it, and avoid introducing potentially problematic precedents into the codebase. There's a couple of days to think about that yet before 1.14 window opens ;-).

ozhuraki commented:
> But mot importantly, there's SYS_LOG_DBG and friends, which can be used directly.

SYS_LOG_* aren’t circumvented here.

> If we let those in, we open a road for "in my special way" for individual programmers

This isn’t "in my special way", but a standard way to do it. SYS_LOG_* are too visually noisy and do not aid code readability.

I am surprised you miss this issue and on the level of pickiness here.

Anyway, I addressed this.

It would be more interesting to see the arguments on why the device tree is needed for this this driver and on needless limitations of IRQ_CONNECT(), since both of them ruin the idea of auto detection in emulators (where this driver is targeting).

pfalcon commented:
@ozhuraki :

> This isn’t "in my special way", but a standard way to do it.

If you grep Zephyr codebase, then you'll see that not in Zephyr.

> SYS_LOG_* are too visually noisy and do not aid code readability.

If it's visually noisy, then it's visually noisy everywhere. Please don't go for "partisan" tactics of patching individual files, go for an RFC to replace it globally in Zephyr, and if ack'ed, go to implement it as such, and then everyone will use something different that SYS_LOG_*. (Indeed, some guy recently did just that. No everyone has to drop everything and convert.) 

> on the level of pickiness here.

That's the impression I'd like to avoid. If we want to develop really high-quality product, we have to agree on some baseline rules and then constrain ourselves to follow them. In personal usage, I have no issues with `D()` and `E()`, but I truly believe they're wrong for a project like Zephyr.

> Anyway, I addressed this.

~~~
#define dev_dbg(fmt, args...) SYS_LOG_DBG(fmt, ## args)
#define dev_err(fmt, args...) SYS_LOG_ERR("Error: " fmt, ## args)
~~~

I don't know why you can't use SYS_LOG_* directly, but I appreciate doing at least something.

Let me +1 this PR, thanks for developing it! But please mind suggestions from @jukkar on squashing fixup commit(s).

pfalcon commented:
@ozhuraki:

I'm not a specialist on these matters, so just 2 cents from me:

> It would be more interesting to see the arguments on why the device tree is needed for this this driver 

Because devicetree is accepted standard for detailed hardware configuration in Zephyr. If this driver did complete PCI autoconfig, it wouldn't be needed. But as it doesn't, it follows standard way of using devicetree. (Thanks for that.)

> and on needless limitations of IRQ_CONNECT(), since both of them ruin the idea of auto detection

My understanding that IRQ_CONNECT() is declarative directive which arranges static, the most efficient IRQ mapping. You understand how it works - it arranges something (ISR address/structure) to be put in ROM, just at the right address.
We also should have way to dynamically set ISR, including allowing drivers to do that. I don't know how much of that "should have" we actually have, we definitely need to elaborate that stuff in Zephyr further.

Note that  device tree was a big step in allowing flexibility of Zephyr hw configuration, even if not *auto*configuration. 

> in emulators (where this driver is targeting).

I won't be surprised if this driver will find wider usages than just for emulators. I won't be surprised even to find a device, where E1000 is used, but without PCI machinery attached. In this case, devicetree support will be right at spot, so thanks again for adding it.

pfalcon commented:
@tbursztyka 

> actually it is "ok" to create new macros for logging on top of SYS_LOG_* but: such should be in capital letters.

It's ok if it's really needed, here the talk was that it's not really needed and about naming.

ozhuraki commented:
@pfalcon
A few key issues have been ignored here. As explained above:

-	There are no non-PCI variants of E1000.
-	This driver primarily targets emulator/simulator cases (qemu, simics) cases. Being able to auto-discover the configuration is a major usability and productivity point here.

> Because devicetree is accepted standard for detailed hardware configuration in Zephyr.

This absolutely doesn’t mean that it should be mindlessly and needlessly applied, disregarding the context and usability.

pfalcon commented:
@ozhuraki, as I told, I'm not a DT expert, mostly try to follow the stuff. So, I hope other folks might answer in more details. But according to my knowledge:

You had (and still have!) in the Kconfig:

~~~
config ETH_E1000_BASE_ADDRESS
	default 0XFEBC0000
~~~

We no longer configure such addresses via Kconfig in Zephyr, only via DT. That's why (per my understanding) you were asked to use DT.

ozhuraki commented:
@pfalcon 
Kconfig is able to override DT and this is still supported.

> That's why (per my understanding) you were asked to use DT.

To me, using it here is a solution looking for a problem and solution in a wrong place that actually introduces a problem.

jukkar commented:
ping @galak, could you re-review the changes so we could get this one merged.

pfalcon commented:
@galak : Ping about approving/merging this. Besides pure usefulness of this driver, it's also the first ethernet driver to use DTS (no such in the tree), so would help with my work on adding smsc9220 support too.

galak reviewed (dismissed):
IRQ stuff looks good, by why do we need the Kconfig for !HAS_DTS_ETHERNET?

ozhuraki commented:
@galak 
> I agree w/Paul, we do we need to support non-DTS Kconfig?

What exactly is still not clear in explanations above on non-DTS use cases?

ozhuraki commented:
@pfalcon 
> Besides pure usefulness of this driver, it's also the first ethernet driver to use DTS (no such in the tree), so would help with my work on adding smsc9220 support too.

Besides, as already explained a few times, irrationally insisting on DTS support here, is making it nearly unusable.

pfl commented:
Could we please get this driver included with the current functionality? There are two other more than capable developers on this thread, @pfalcon and @galak, that can come with the additional patches, especially since there aren't any non-PCI hw configurations available.

galak commented:
> Besides, as already explained a few times, irrationally insisting on DTS support here, is making it nearly unusable.

The project as a whole is using device tree for such information and moving away from Kconfig.  The idea of adding new driver support and still providing Kconfig options is against the direction the project is moving in.

Maybe I'm misunderstanding something, but I'm not seeing what the Kconfig support is providing that we can't do with device tree for the SoC that is being supported at this point.

nashif reviewed (changes requested):
agree with @galak, we should be able to move the base address Kconfig to DTS, if we need to change that for some specific platform configuration, we can use overlays or fixups.

ozhuraki commented:
@nashif 
> agree with @galak, we should be able to move the base address Kconfig to DTS

It’s actually already in DTS.

> […] if we need to change that for some specific platform configuration, we can use overlays or fixups.

Kconfig here does exactly that.

pfalcon commented:
> other [...] developers [...] that can come with the additional patches,

Exactly my attitude, we rarely get everything perfect from the first time, and I +1'ed from my side. But I'm just a developer, and just learn about DT usage in Zephyr, as many other folks too.

galak commented:
> Kconfig here does exactly that.

the point is that we are trying to deprecated Kconfig usage for this purpose and thus dts overlay is the way to do that right now until PCI bus scanning & irq setup is supported for your needs.

Thus the desire to remove the Kconfig bits.

nashif commented:
> Kconfig here does exactly that.

ok, then I am confused. Why do you carry the same configuration in two places? Why don't we just move to DTS and skip the kconfig part?

ozhuraki commented:
@galak 
> Maybe I'm misunderstanding something, but I'm not seeing what the Kconfig support is providing that we can't do with device tree for the SoC that is being supported at this point.

I think there’s lack of understanding that the base address even on one SOC can change due to PCI enumeration/assignment and might depend on what PCI devices are present.

In case of emulators/simulators (where this driver is targeting), this might even depend on the order of PCI device instantiation (for example, in qemu).

Requesting DT here is solving the wrong problem in a wrong place. 

And, without an ability to override, this DT only solution will be just unusable.

nashif commented:
> And, without an ability to override, this DT only solution will be just unusable.

you can override using overlays or fixup files. see https://docs.zephyrproject.org/latest/application/application.html#application-dt

ozhuraki commented:
@nashif 
> you can override using overlays or fixup files. see https://docs.zephyrproject.org/latest/application/application.html#application-dt

OK, thanks for the input. If Kconfig is deprecated for DT overrides, let’s use this.

pfalcon commented:
@ozhuraki : Did you have a chance to look into this?

nashif commented:
@ozhuraki ping

ozhuraki commented:
@pfalcon @nashif 
Removed Kconfig entries for the base address and IRQ. Please take a look.

pfalcon commented:
@ozhuraki : Thanks for making changing and squashing everything related to eth_e1000.c into one commit.

@galak, @nashif : Ping for re-review.

galak reviewed (commented):
Looks good to me.  I noticed some logging support, not sure if that needs updating for the new logging subsystem.

ozhuraki commented:
@galak 
> I noticed some logging support, not sure if that needs updating for the new logging subsystem.

Forgot to mention: I also updated it to use the new logging.

pfalcon commented:
So, just as with https://github.com/zephyrproject-rtos/zephyr/pull/10097, I'm trying to "reverse engineer" how to get this to run. But unlike the above patch, I failed. My steps:

1. Enable the e1000 driver:

~~~
+++ b/samples/net/echo_server/prj.conf
+CONFIG_PCI=y
+CONFIG_PCI_ENUMERATION=y
+CONFIG_ETH_E1000=y
+CONFIG_NET_SLIP_TAP=n
+CONFIG_NET_L2_ETHERNET=y
~~~ 

Start with:

~~~
sudo /home/pfalcon/opt/zephyr-sdk-0.9.3/sysroots/x86_64-pokysdk-linux/usr/bin/qemu-system-i386 \
-m 8 -cpu qemu32,+nx,+pae -device isa-debug-exit,iobase=0xf4,iosize=0x04 \
-no-reboot -nographic -vga none -display none \
-clock dynticks -no-acpi -balloon none \
-machine type=pc-0.14 -pidfile qemu.pid \
-serial mon:stdio \
-netdev tap,id=n1 \
-device e1000,netdev=n1 \
-object filter-dump,id=f1,netdev=n1,file=dump.dat \
-kernel /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/samples/net/echo_server/outdir/qemu_x86/zephyr/zephyr.elf
~~~

What I get on Zephyr console is:
~~~
W: /etc/qemu-ifup: no bridge for guest interface found
qemu-system-i386: warning: Unknown firmware file in legacy mode: genroms/multiboot.bin

e1000: TDH wraparound @0, TDT 1, TDLEN 0
e1000: TDH wraparound @0, TDT 1, TDLEN 0
***** Booting Zephyr OS zephyr-v1.13.0-935-gb307be0382 *****
Zephyr Shell, Zephyr version: 1.13.99
Type 'help' for a list of available commands
shell> e1000: TDH wraparound @0, TDT 1, TDLEN 0
e1000: TDH wraparound @0, TDT 1, TDLEN 0
e1000: TDH wraparound @0, TDT 1, TDLEN 0
~~~

Pings to 192.0.2.1 don't work, per Wireshark, there're no replies to ARP, "net stats" shows:

~~~
Global statistics
=================
IPv6 recv      0	sent	5	drop	0	forwarded	0
IPv6 ND recv   0	sent	5	drop	0
IPv6 MLD recv  0	sent	0	drop	0
IPv4 recv      0	sent	0	drop	0	forwarded	0
IP vhlerr      0	hblener	0	lblener	0
IP fragerr     0	chkerr	0	protoer	0
ICMP recv      0	sent	0	drop	0
ICMP typeer    0	chkerr	0
UDP recv       0	sent	0	drop	0
UDP chkerr     0
TCP bytes recv 0	sent	0
TCP seg recv   0	sent	0	drop	0
TCP seg resent 0	chkerr	0	ackerr	0
TCP seg rsterr 0	rst	0	re-xmit	0
TCP conn drop  0	connrst	0
Bytes received 0
Bytes sent     288
Processing err 0
~~~

pfalcon commented:
Result of PCI detection:
~~~
0:2 8086:100e class: 0x2, 0, 0, MEM,addrs: 0xfebc0000-0xfebdffff, IRQ 11
~~~

IRQ matches what's hardcoded in the DTS, but I don't see the driver ISR firing.

pfalcon commented:
> e1000: TDH wraparound @0, TDT 1, TDLEN 0

This message doesn't appear eth_e1000.c, and it was an unmonotonic thought to imagine it may come from QEMU, and indeed it is, as mentioned e.g. here: https://pdos.csail.mit.edu/6.828/2016/labs/lab6/

pfalcon commented:
> e1000: TDH wraparound @0, TDT 1, TDLEN 0

Well, these appear to be coming before e1000_init() is called, none after it.

ozhuraki commented:
@pfalcon

Looks to me you got everything almost right, just sort out the details with setting the tap interface.

A quick way to check if problems are tap related, is something like this:

qemu ... -device e1000,netdev=net1 -netdev socket,id=net1,udp=localhost:port1,localaddr=localhost:port2

and then look with the wireshark.

To smoke test the receive path this way, net-test-suites would work almost out of the box.

It might be easier to troubleshooting this with the static IP address configuration of your app, enable the debug output for the driver, get the signal through with the outgoing traffic, and then incoming.

```
e1000: TDH wraparound @0, TDT 1, TDLEN 0
```

Our usage is legal here. That warning is extra and, perhaps, needs a patch to qemu.

pfalcon commented:
> /home/pfalcon/opt/zephyr-sdk-0.9.3/sysroots/x86_64-pokysdk-linux/usr/bin/qemu-system-i386

I also tried with Zephyr SDK 0.9.4-rc3. With it, iPXE BIOS (or whatever it's called) is visibly booting, and there're no "e1000: TDH wraparound @0, TDT 1, TDLEN 0" warnings.

The rest of behavior is the same - pings don't go thru, IRQs not firing.

pfalcon commented:
So, looking thru the captured traffic (`-object filter-dump,id=f1,netdev=n1,file=dump.dat` in my command above), seems that the e1000 manages to send out something which Wireshark on Linux side sees. So again, the reverse direction doesn't work, starting with getting a RX interrupt. 
 
~~~
4	1.352135	::	ff02::1:ff12:3456	ICMPv6	78		Neighbor Solicitation for fe80::5054:ff:fe12:3456
6	1.355723	::	ff02::1:ff00:1	ICMPv6	78		Neighbor Solicitation for 2001:db8::1
~~~

pfalcon commented:
@ozhuraki:

> A quick way to check if problems are tap related, is something like this:
> qemu ... -device e1000,netdev=net1 -netdev socket,id=net1,udp=localhost:port1,localaddr=localhost:port2

As I mentioned, I initially tried a similar approach with https://github.com/zephyrproject-rtos/zephyr/pull/10097, where everything worked, so there're no suspicions re: TAP interface.

I also initially started with "QEMU SLIRP" networking https://wiki.qemu.org/Documentation/Networking#User_Networking_.28SLIRP.29 , but that didn't work and that's shy I switched to TAP.

So, I'm not sure what your expectations re: the commandline params you gave, but mine were fully satisfied when I used `-netdev socket,id=n1,udp=localhost:4242,localaddr=localhost:4242`, then `nc -u localhost 4242` on Linux side. So, for Zephyr to respond to something, it first need to resolve something's address via ARP, and of course, that can't work over "socket,id=n1,udp", so in Wireshark, I got the expected ICMP Destination unreachable.

And still not RX interrupts. (I have 
~~~
 static void e1000_isr(struct device *device)
 {
+printk("e1000_isr\n");
~~~
obviously).

pfalcon commented:
@ozhuraki : Can we go the other way, can you please provide the exact commands showing the complete IP communication cycle (e.g, ping running), which work for you with the latest master?

pfalcon commented:
@jukkar : What about you, does this driver work for you?

jukkar commented:
I used this command line
```
qemu-system-i386 -m 8 \
      -cpu qemu32,+nx,+pae -device isa-debug-exit,iobase=0xf4,iosize=0x04 \
      -no-reboot -nographic -machine type=pc \
      -display none -clock dynticks -no-acpi -balloon none \
      -pidfile qemu.pid \
      -serial mon:stdio \
      -net nic,model=e1000,netdev=znet \
      -netdev tap,id=znet,script=no,downscript=no,ifname=znet \
      -kernel zephyr/zephyr.elf
```

pfalcon commented:
@jukkar: My system qemu:

~~~
$ qemu-system-i386 --version
QEMU emulator version 2.11.1(Debian 1:2.11+dfsg-1ubuntu7.5)
Copyright (c) 2003-2017 Fabrice Bellard and the QEMU Project developers
~~~

So, when I tried to run the command as provided by you, I got complains about various "bios" files missing, e.g. efi-e1000.rom, so I picked it/them up from internets/local hdd. 

Then it started to start.

I also changed -netdev line to:

~~~
-netdev tap,id=znet
~~~

so it created tap0 for me, which I `ifconfig tap0 192.0.2.2`'ed. Then the situation is absolutely the same as with my previous attempts - pings don't work, the driver doesn't see a single packet, I assume RX interrupts don't fire as before.

This is a cute situation as in https://github.com/zephyrproject-rtos/zephyr/issues/8869 , where everything works for me, but doesn't work for @nniranjhana and @nashif . Except we here try to call qemu versions and commandlines, but guys there don't provide any additional info.

jukkar commented:
I installed zephyr-sdk_0.9.4-rc3, it is missing some rom files but those can be temporarily picked from host system. With that qemu the ping works for about 6 seconds, after which the network connectivity stops.

What is also weird that I have now same symptoms with host qemu (connectivity stops after 6 seconds), I really wonder what has happened because this connectivity worked ok yesterday when I was testing it.
